### PR TITLE
[WIP] Cancellation support

### PR DESCRIPTION
**Pull request by ulugbekna**

### CODE REVIEW

**Comment by ulugbekna** [comment]
The last commit is a step forward in cancelling requests in jsonrpc, but it's not working (tbh don't know why yet) - the request is not cancelled.

File: lsp/src/client_notification.ml [hunk 1]
```diff
@@ -30,6 +32,7 @@ let yojson_of_t = function
   | TextDocumentDidChange params ->
     DidChangeTextDocumentParams.yojson_of_t params
   | TextDocumentDidClose params -> DidCloseTextDocumentParams.yojson_of_t params
+  | CancelRequest params -> CancelParams.yojson_of_t params
```
Comment (rgrinberg) [inline lsp/src/client_notification.ml:35]: The changes introducing `CancelRequest` can be merged.

File: lsp/src/jsonrpc.mli [hunk 2]
```diff
@@ -135,4 +135,6 @@ end) : sig
   val notification : _ t -> Message.notification -> unit Fiber.t
 
   val request : _ t -> Message.request -> Response.t Fiber.t
+
+  val try_cancel_request : _ t -> [ `Cancelled | `Failed ] Fiber.t
```
Comment (rgrinberg) [inline lsp/src/jsonrpc.mli:139]: I'm not sure that this is the right API. We should be able to ask for a cancellation on an arbitrary request (given the Id).

File: lsp/src/client_notification.ml (hunk 1 above)
Comment (rgrinberg) [reply lsp/src/client_notification.ml:35]: I cherry picked this into #322

**Pull request closed without merging**